### ISSUE TITLE
raft: decouple raft release from etcd release

### ISSUE THREAD

**Issue by ahrtr** [etcd-io/etcd#14713]
### What would you like to be added?

Currently the [raft](https://github.com/etcd-io/etcd/tree/main/raft) is released & tagged together with etcd. For example, when we release etcd 3.5.5, we will create tag `v3.5.5` for etcd and `raft/v3.5.5` for raft at the same time respectively. 

raft is an important & stable module, it may have different evolution pace from etcd. I think it would be better to qualify and release raft separately, just like what we did for [BoltDB](https://github.com/etcd-io/bbolt). 

We probably need to change the module name from `go.etcd.io/etcd/raft/v3` to `go.etcd.io/raft`. Since it's a new beginning, we can release raft starting from `v1`, which isn't needed in the module name. We need to add the version into the module name starting from `v2`, namely `go.etcd.io/raft/v2`.  FYI. [Module version numbering](https://go.dev/doc/modules/version-numbers)

We can even create a separate repository for raft, such as `github.com/etcd-io/raft`. But it isn't a blocker for this request, either way we will release raft separately. 

We need to agree on the release criteria of raft if we follow this approach. For the first step, I think we can keep doing what we already have,
1. The [datadriven test](https://github.com/etcd-io/etcd/tree/main/raft/rafttest);
2. etcd [workflows](https://github.com/etcd-io/etcd/tree/main/.github/workflows);
3. Kubernetes SIG Scalability test. (Note: @serathius maintains etcd version in Kubernetes)

Since [release-3.5](https://github.com/etcd-io/etcd/tree/release-3.5) is a stable release, so we will not change anything on 3.5 and [raft in 3.5](https://github.com/etcd-io/etcd/tree/release-3.5/raft). This change is only for the `main` branch. 

## So in summary, the request has three points:
1. release raft separately;
2. rename raft module name from `go.etcd.io/etcd/raft/v3` to `go.etcd.io/raft`.
3. (optional) create a separate repository (e.g github.com/etcd-io/raft) for raft. 

@hexfusion @mitake @ptabor @serathius @spzala @tbg @nvanbenschoten @pavelkalinnikov Please share your thoughts. thx. 

### Why is this needed?

see above.

**Comment by serathius** [comment]
Moving raft to separate repo was already discussed and got early support in etcd maintainer mailing list. Overall it make sense for me as it will allow raft maintainers to implement their changes without being constrained by etcd. 

My main worry is about investment cost. Will investing time and effort into code organization is as important now if we could invest it in other places.  I think it's up to @tbg to decide if splitting raft would be a net positive investment.

**Comment by tbg** [comment]
> I think it's up to @tbg to decide if splitting raft would be a net positive investment.

The main goal for now should be making it possible for `etcd` to release with a version of raft of their choosing, i.e. establishing symmetry between `etcd` and the other consumers of `raft` (like CockroachDB). Changes to raft should not have to grind to a halt, for example, in the lead-up to an etcd release but this tends to be prudent if both are to be released from a joint SHA.

For the community having more frequent releases of raft could be nice; at CockroachDB we don't depend on releases; we have a good handle on all changes to `raft` and pick a SHA directly. I am unsure how many other users are out there that would find more frequent releases useful but they definitely wouldn't hurt.

I don't consider it a goal in and of itself to move `raft` out into a separate repository. This would have some benefits - for example, CI is often flaky and rarely is it ever due to a change made in a PR to `raft` - but at the same time the additional coverage we're getting through `etcd` is nice, especially since it involves end-to-end testing of clusters containing different versions of raft.[^1]


Is it even practical to release two projects out of the same repo? I don't think the Github releases are able to express this very well. Won't they just tar up the contents of the repo as of a single SHA? Doesn't go.mod use git tags to determine which versions exist? So I wonder if just separating the release cycles in practice forces us into a separate repo. I might be missing something that makes it possible to avoid it.

If we go for a separate repo, I think a standard set of Github actions (unit tests, vet, staticcheck) is sufficient - `raft` is a very vanilla Go library. We can start with the same maintainer sets; I see no reason to split off from `etcd` any more than necessary by decoupled releases.

[^1]: But even in the same repo, with releases tagged separately, the `etcd` tests would be using whatever version of raft they pinned, and thus wouldn't exercise the `raft` code touched in a PR. Even worse, developers might mistakenly think this would still be the case. We could fix this, but then we'd be ~doubling the test suite size, which doesn't seem useful.

**Comment by pavelkalinnikov** [comment]
FWIW, last time I checked it was possible to release different modules independently within the same repo. The magic is in prefixing the tag with the submodule name/path within the repo. For example, see "Tag based versioning" [here](https://stackoverflow.com/questions/55041915/multiple-modules-within-the-same-project#answer-71984158).

So, you could have `v3.x` tags for most of the etcd stuff, and in addition `raft/v1.0` for raft. Go module resolution would also work.

**Comment by pavelkalinnikov** [comment]
If we end up "restarting" the versioning for the raft package, I would suggest not rushing into `v1` straight away though. Stay in `v0` for a bit, reevaluate the API. Once in `v1`, there is a stricter semver expectation at the code level (not only at the behaviour). Systems usually ignore this rule, but `etcd/raft` is more at the "library" end, so code-level semver would be ideal.

**Comment by tbg** [comment]
> FWIW, last time I checked it was possible to release different modules independently within the same repo. The magic is in prefixing the tag with the submodule name/path within the repo. For example, see "Tag based versioning" [here](https://stackoverflow.com/questions/55041915/multiple-modules-within-the-same-project#answer-71984158), and some suggestions in Go [docs](https://go.dev/doc/modules/managing-source#multiple-module-source).

@pavelkalinnikov right this is Go, good to know, but etcd also relies on Github releases: https://github.com/etcd-io/etcd/releases
I don't think there is a way to say "make a release but only of this directory". So awkwardly, an etcd release will contain raft code that isn't actually used in that release. And more awkwardly, a raft release is really a full etcd tarball as well. I don't see a way around that and (not knowing any better) think that the "Github way" of solving this is to "just" have a separate repo.

**Comment by tbg** [comment]
> If we end up "restarting" the versioning for the raft module, I would suggest not rushing into v1 straight away though. Use the opportunity and stay in v0 for a bit, reevaluate the API. 

`raft` is a mature library that is (presumably) used in lots of places, so we would want to have continuity with the current version, i.e. have a production-ready release right away to make it easy for users to switch to. I think it would raise questions to go to v0, or even to v1. Currently we are at v3 and so shouldn't we continue there? I don't see a better way to make clear that the standalone repo is the "latest and greatest". We can write that into docs 100 times, but versions are supposed to be the ground truth.

I understand the desire to evolve the `raft` API and I am in favor of doing that, but only in a way that adds ~zero friction to existing users. So I would be hesitant to use the repo switch as a reason to make API changes. Besides, I think most API changes we could make would be better done in the same module version, but by providing a new implementation point.

>  (not only at the behaviour; e.g. simple renames of exported fields are technically breaking changes)

It's a good idea to point out our policy about breaking changes. I would like us to be able to occasionally rename a field, move a struct, etc, and we should be clear on whether that means minting a new major release or not. I tend to think it would be nice to avoid it (don't see the cost-benefit analysis playing out favorably) but open to a discussion. Pavel, since you seem to a) know a lot about versioning and b) seem to care, maybe you could continue leading that discussion?

**Comment by ahrtr** [comment]
> I think it's up to @tbg to decide if splitting raft would be a net positive investment.

I don't agree this. This is a big structure change, nobody can make any unilateral decision. I think it should be up to all existing etcd maintainers' decision, and of course the feedback from the community. If all etcd maintainers reach an agreement, then we should send a discussion topic in etcd-dev to get feedback as well. 

Splitting raft has impact on all raft users, but I think it should be positive impact, because raft can be released & maintained separately. etcd doesn't necessarily always to depend on the latest raft code. Picking a specific commit can also work, but it makes more sense to depend on tags.

cc @hexfusion @mitake @ptabor @spzala could you please share your thoughts as well? thx. 


> I think it would raise questions to go to v0, or even to v1. Currently we are at v3 and so shouldn't we continue there?

Agreed. It makes sense to me. 



> I don't think there is a way to say "make a release but only of this directory". So awkwardly, an etcd release will contain raft code that isn't actually used in that release. And more awkwardly, a raft release is really a full etcd tarball as well.

The best approach seems to be the third point I mentioned above: create a separate repository for raft, such as `github.com/etcd-io/raft`.  Sounds good?

**Comment by ahrtr** [comment]
> We can start with the same maintainer sets;

Fair enough to me.

**Comment by tbg** [comment]
> The best approach seems to be the third point I mentioned above: create a separate repository for raft, such as github.com/etcd-io/raft. Sounds good?

That's the only clean way I can think of, given the limitations of a Github-supported release process.

**Comment by tbg** [comment]
> > We can start with the same maintainer sets;
> 
> Fair enough to me.

Just an observation: I'm noticing that our @etcd-io/maintainers-etcd  GH team is anchored on the etcd-io org, so hopefully this wouldn't even require maintaining parallel worlds.

**Comment by mitake** [comment]
I think having a separated raft module is a good idea. Various users of the module can have their own suitable way for controlling risk and benefit from new improvements.

>> So awkwardly, an etcd release will contain raft code that isn't actually used in that release. And more awkwardly, a raft release is really a full etcd tarball as well.

> The best approach seems to be the third point I mentioned above: create a separate repository for raft, such as github.com/etcd-io/raft. Sounds good?

I agree with @ahrtr . Having raft code that isn't actually used in that release seems to be hard to manage. Separating repositories seems to be good.

**Comment by spzala** [comment]
+1 Any of the three approaches seems good but I think for the long term benefits`github.com/etcd-io/raft` seems a better approach. We should however make sure that we are good with etcd test stability before we start working on it. Thanks @ahrtr  for initiating this discussion.

**Comment by ahrtr** [comment]
Thanks  @mitake and  @spzala for the feedback.

> Just an observation: I'm noticing that our @etcd-io/maintainers-etcd GH team is anchored on the etcd-io org, so hopefully this wouldn't even require maintaining parallel worlds.

I do not quite catch the point probably due to my English(not a native speaker:( ). My understanding the `etcd-io/maintainers-etcd ` team is supposed to only maintain [etcd](https://github.com/etcd-io/etcd/). And we will add all the existing members into the maintainer list of the new raft repo `github.com/etcd-io/raft `.

**Comment by tbg** [comment]
I was saying that we could keep the `etcd-maintainers` team in charge of raft initially but you are probably right that if we're separating it we might as well separate it for good. I certainly have no business maintaining etcd once raft is removed from it, since I haven't contributed to it at all.

**Comment by ahrtr** [comment]
cc @dims @liggitt @deads2k @lavalamp for awareness. 

Please note that the change should have no impact on Kubernetes at all.

**Comment by ahrtr** [comment]
The plan for the next step is:
1. Create a separate repo (`github.com/etcd-io/raft`) for raft. (See my PoC https://github.com/ahrtr/raft);
2. Change the module name from `go.etcd.io/etcd/raft/v3` to `go.etcd.io/raft/v3`;
3. Since [raft/v3.6.0-alpha.0](https://github.com/etcd-io/etcd/releases/tag/raft%2Fv3.6.0-alpha.0) is already out, so the first raft version will be v3.6.0. 
4. Create a new github team `maintainers-raft`, and add all the existing etcd maintainers (@ahrtr @hexfusion @mitake @ptabor @serathius @spzala @tbg ) into the new team.
5. Update etcd to point to the new raft module.
6. Once everything is working well, remove the [raft](https://github.com/etcd-io/etcd/tree/main/raft) directory from etcd repo.

Since we do not get feedback from @ptabor and @hexfusion yet, so I will send an email to them (cc etcd-maintainers) as well. Once all the existing maintainers agree on the approach, then I will post a message in [etcd-dev](https://groups.google.com/g/etcd-dev) to collect feedback.

Please kindly feedback if I miss anything of you have any comment or concerns.

**Comment by hexfusion** [comment]
This in general makes sense to me as the raft package is widely used outside of etcd. As far as cycles to implement I would defer decision to more active maintainers.

**Comment by ptabor** [comment]
Make sense to me.

**Comment by ahrtr** [comment]
Thanks both @hexfusion and @ptabor for the feedback. 

**So we have got positive feedback from all etcd maintainers**. I will post a message to [etcd-dev](https://groups.google.com/g/etcd-dev) to collect feedback as well.

**Comment by ahrtr** [comment]
FYI. https://groups.google.com/g/etcd-dev/c/PC1-UXYiWwk

**Comment by ptabor** [comment]
I think we should also migrate: 
  https://github.com/etcd-io/etcd/tree/main/contrib/raftexample

It seems to have only a dependency on etcd Snapshot,
but I think it should get decoupled from etcd even if it means cloning the Snapshot proto
to make it part of learning experience for the RAFT lib.

**Comment by ahrtr** [comment]
> I think we should also migrate:
> https://github.com/etcd-io/etcd/tree/main/contrib/raftexample

Makes sense to me, and thx @ptabor for explicitly raised this. Previously I thought about this. It isn't a big deal and nobody explicitly mentioned it, so I did not do it in my PoC. 



> It seems to have only a dependency on etcd Snapshot,
> but I think it should get decoupled from etcd even if it means cloning the Snapshot proto
> to make it part of learning experience for the RAFT lib.

Let me take care of this. The raft example shouldn't depend on etcd at all.

**Comment by ahrtr** [comment]
> It seems to have only a dependency on etcd Snapshot,
> but I think it should get decoupled from etcd even if it means cloning the Snapshot proto
> to make it part of learning experience for the RAFT lib.

I just had a quick evaluation on this, and it seems that it isn't that simple. The raft implementation follows a minimalistic design philosophy, and it depends on the [rafthttp](https://github.com/etcd-io/etcd/tree/main/server/etcdserver/api/rafthttp) and [wal](https://github.com/etcd-io/etcd/tree/main/server/storage/wal) as well. So I regard the [raftexample](https://github.com/etcd-io/etcd/tree/main/contrib/raftexample) as an example specific to etcd. So I will not migrate it to the new repo planned for raft. Please let me know if you have any concern. 


Note: I am still waiting for the feedback from https://groups.google.com/g/etcd-dev/c/PC1-UXYiWwk before taking any action.

**Comment by Elbehery** [comment]
+1

**Comment by ptabor** [comment]
> > It seems to have only a dependency on etcd Snapshot,
> > but I think it should get decoupled from etcd even if it means cloning the Snapshot proto
> > to make it part of learning experience for the RAFT lib.
> 
> I just had a quick evaluation on this, and it seems that it isn't that simple. The raft implementation follows a minimalistic design philosophy, and it depends on the [rafthttp](https://github.com/etcd-io/etcd/tree/main/server/etcdserver/api/rafthttp) and [wal](https://github.com/etcd-io/etcd/tree/main/server/storage/wal) as well. So I regard the [raftexample](https://github.com/etcd-io/etcd/tree/main/contrib/raftexample) as an example specific to etcd. So I will not migrate it to the new repo planned for raft. Please let me know if you have any concern.

Thank you for checking. Makes sense to me. 
I would consider then just removing this contrib eventually. Etcd maintainers does not need 'etcd light` that proven to have bugs on its own. Contributing (migrating to) self-contained raft-minimal example might be interesting low-priority starter project.

**Comment by ahrtr** [comment]
Thanks @Elbehery for the feedback.

> I would consider then just removing this contrib eventually. Etcd maintainers does not need 'etcd light` that proven to have bugs on its own.

Thanks for the response. Are you suggesting to remove [raftexample](https://github.com/etcd-io/etcd/tree/main/contrib/raftexample) from etcd repo eventually? I have no objection for it; but I would suggest to keep it as it's for now, until we finish migrating raft to a separate repo and implemented a self-contained raft example in the raft repo.  



> Contributing (migrating to) self-contained raft-minimal example might be interesting low-priority starter project.

Yes, we can create a ticket to track this in the new raft project once it's available.

**Comment by ahrtr** [comment]
All the feedback we got so far is positive, so I will kick off the migration process per https://github.com/etcd-io/etcd/issues/14713#issuecomment-1320385870 in the following week. Please let me know if anyone has any concern.

**Comment by ahrtr** [comment]
Update: the new repo [raft](https://github.com/etcd-io/raft) is already created, and raised the first PR https://github.com/etcd-io/raft/pull/1 .

**Comment by ahrtr** [comment]
The migration is done. The raft was moved to a separate repo https://github.com/etcd-io/raft, and the module name was renamed to [go.etcd.io/raft/v3](http://go.etcd.io/raft/v3).

Thanks all for supporting this.

**Comment by ahrtr** [comment]
Also sent notification in https://groups.google.com/g/etcd-dev/c/PC1-UXYiWwk
